File: go/internal/numfmt/numfmt.go
// Package numfmt formats an IEEE-754 double exactly as ECMAScript's
// Number::toString does (SPEC §6; conformance/README.md): shortest round-trip
// decimal, integer-valued doubles without a decimal point, fixed notation
// inside the standard range and exponential outside it.
package numfmt

import (
	"math"
	"strconv"
	"strings"
)

// Format renders f using ECMAScript Number-to-String semantics.
func Format(f float64) string {
	if f == 0 {
		return "0" // covers +0 and -0
	}
	if math.IsNaN(f) {
		return "NaN"
	}
	if math.IsInf(f, 1) {
		return "Infinity"
	}
	if math.IsInf(f, -1) {
		return "-Infinity"
	}

	sign := ""
	if f < 0 {
		sign = "-"
		f = -f
	}

	// Shortest round-trip digits + decimal exponent via the 'e' form.
	e := strconv.FormatFloat(f, 'e', -1, 64) // e.g. "3.771e-01"
	mantissa, expStr, _ := strings.Cut(e, "e")
	digits := strings.Replace(mantissa, ".", "", 1)
	exp, _ := strconv.Atoi(expStr) // power of ten of the first digit
	k := len(digits)               // significant digit count
	n := exp + 1                   // decimal point position (ES §Number::toString)

	switch {
	case k <= n && n <= 21:
		return sign + digits + strings.Repeat("0", n-k)
	case 0 < n && n <= 21:
		return sign + digits[:n] + "." + digits[n:]
	case -6 < n && n <= 0:
		return sign + "0." + strings.Repeat("0", -n) + digits
	default:
		return sign + exponential(digits, k, n)
	}
}

func exponential(digits string, k, n int) string {
	head := digits[:1]
	if k > 1 {
		head += "." + digits[1:]
	}
	e := n - 1
	if e >= 0 {
		return head + "e+" + strconv.Itoa(e)
	}
	return head + "e-" + strconv.Itoa(-e)
}

File: python/src/gridmd/numfmt.py
"""ECMAScript ``Number → String`` formatting.

The canonical dump (``conformance/README.md``) renders IEEE-754 doubles exactly
as JavaScript's ``String(n)`` does: shortest round-trip decimal, integer-valued
doubles with no decimal point (``1000`` not ``1e3`` not ``1000.0``), fixed
notation inside the standard range and exponential outside it. Python's
``repr``/``str`` disagree on both integers (``1000.0``) and exponent placement
(``1e+21``/``1e-07``), so this is hand-rolled to match V8 exactly.
"""

from __future__ import annotations

import math
from decimal import Decimal


def format_number(value: float | int) -> str:
    """Render ``value`` using ECMAScript ``Number::toString`` semantics."""
    f = float(value)
    if f == 0:
        return "0"  # covers +0 and -0
    if math.isnan(f):
        return "NaN"
    if math.isinf(f):
        return "Infinity" if f > 0 else "-Infinity"

    sign = ""
    if f < 0:
        sign = "-"
        f = -f

    # Shortest round-trip significant digits + decimal exponent.  repr() gives
    # the shortest decimal that round-trips; Decimal(...).normalize() strips
    # trailing zeros so the remaining digit tuple is exactly the significand.
    digits_tuple, exp = _significand(f)
    digits = "".join(str(d) for d in digits_tuple)
    k = len(digits)  # significant digit count
    n = k + exp  # ES §Number::toString decimal-point position

    if k <= n <= 21:
        return sign + digits + "0" * (n - k)
    if 0 < n <= 21:
        return sign + digits[:n] + "." + digits[n:]
    if -6 < n <= 0:
        return sign + "0." + "0" * (-n) + digits
    return sign + _exponential(digits, k, n)


def _significand(f: float) -> tuple[tuple[int, ...], int]:
    """Return ``(digit_tuple, exp)`` such that ``value == int(digits) * 10**exp``
    with no trailing zero digits."""
    _, digits_tuple, exp = Decimal(repr(f)).normalize().as_tuple()
    return digits_tuple, exp


def _exponential(digits: str, k: int, n: int) -> str:
    head = digits[0]
    if k > 1:
        head += "." + digits[1:]
    e = n - 1
    return head + ("e+" + str(e) if e >= 0 else "e-" + str(-e))
